File: services/ai-sidecar/src/application/services/dispatch/dispatch_analytics_service.py
"""Dispatch operations analytics service."""

from __future__ import annotations

from collections import defaultdict
from collections.abc import Iterable
from datetime import datetime, timedelta
from math import sqrt
from typing import Any

from src.domain.utils.time_utils import utc_now


class DispatchAnalyticsService:
    """Aggregate dispatch operations KPIs within a time window."""

    REPLAN_ACTION_PREFIXES = ("replanned",)

    def __init__(
        self,
        order_repo: Any,
        conflict_service: Any,
        resource_utilization_service: Any,
    ) -> None:
        self._order_repo = order_repo
        self._conflict_service = conflict_service
        self._resource_utilization_service = resource_utilization_service

    async def get_operations_summary(
        self,
        window_start: datetime | None = None,
        window_end: datetime | None = None,
    ) -> dict[str, Any]:
        ws, we = self._resolve_window(window_start, window_end)
        orders = await self._get_orders(ws, we)
        conflict_items = await self._get_conflicts(ws, we, len(orders))
        conflict_order_ids = self._conflict_order_ids(conflict_items)
        replan_order_ids = await self._get_replanned_order_ids(orders)
        response_minutes = self._response_minutes(orders)
        team_loads = self._team_occupied_minutes(orders, ws, we)
        equipment_utilization = await self._resource_utilization_service.get_equipment_utilization(ws, we)
        key_stats = self._key_order_stats(orders)
        assigned_orders = [order for order in orders if self._is_assigned(order)]
        completed_orders = [order for order in orders if self._status_value(order) == "completed"]

        equipment_idle_rate = 0.0
        if equipment_utilization:
            avg_rate = sum(float(item.get("utilization_rate") or 0.0) for item in equipment_utilization) / len(
                equipment_utilization
            )
            equipment_idle_rate = round(max(0.0, 1.0 - avg_rate), 4)

        denominator = max(1, len(assigned_orders))
        return {
            "window_start": ws.isoformat(),
            "window_end": we.isoformat(),
            "assigned_order_count": len(assigned_orders),
            "completed_order_count": len(completed_orders),
            "conflict_count": len(conflict_items),
            "conflict_order_count": len(conflict_order_ids),
            "conflict_rate": round(len(conflict_order_ids) / denominator, 4),
            "replanned_order_count": len(replan_order_ids),
            "replan_rate": round(len(replan_order_ids) / denominator, 4),
            "avg_dispatch_response_minutes": self._average(response_minutes),
            "team_load_balance_score": self._load_balance_score(team_loads.values()),
            "equipment_idle_rate": equipment_idle_rate,
            "key_order_count": key_stats["key_order_count"],
            "key_order_ontime_rate": key_stats["key_order_ontime_rate"],
        }

    async def get_breakdown(
        self,
        *,
        window_start: datetime | None = None,
        window_end: datetime | None = None,
        group_by: str = "team",
    ) -> list[dict[str, Any]]:
        ws, we = self._resolve_window(window_start, window_end)
        orders = await self._get_orders(ws, we)
        conflict_items = await self._get_conflicts(ws, we, len(orders))
        conflict_order_ids = self._conflict_order_ids(conflict_items)
        replan_order_ids = await self._get_replanned_order_ids(orders)

        groups: dict[str, dict[str, Any]] = {}
        for order in orders:
            key, label = self._group_key(order, group_by)
            entry = groups.setdefault(
                key,
                {
                    "group_key": key,
                    "group_label": label,
                    "order_count": 0,
                    "assigned_order_count": 0,
                    "completed_order_count": 0,
                    "occupied_minutes": 0.0,
                    "conflict_order_count": 0,
                    "replanned_order_count": 0,
                    "response_minutes": [],
                },
            )
            entry["order_count"] += 1
            if self._is_assigned(order):
                entry["assigned_order_count"] += 1
            if self._status_value(order) == "completed":
                entry["completed_order_count"] += 1
            entry["occupied_minutes"] += self._occupied_minutes(order, ws, we)
            if str(getattr(order, "id", "")).strip() in conflict_order_ids:
                entry["conflict_order_count"] += 1
            if str(getattr(order, "id", "")).strip() in replan_order_ids:
                entry["replanned_order_count"] += 1
            response_minutes = self._response_minutes_for_order(order)
            if response_minutes is not None:
                entry["response_minutes"].append(response_minutes)

        items: list[dict[str, Any]] = []
        for item in groups.values():
            assigned_count = max(1, int(item["assigned_order_count"]))
            items.append(
                {
                    "group_key": item["group_key"],
                    "group_label": item["group_label"],
                    "order_count": item["order_count"],
                    "assigned_order_count": item["assigned_order_count"],
                    "completed_order_count": item["completed_order_count"],
                    "occupied_minutes": round(float(item["occupied_minutes"]), 2),
                    "conflict_order_count": item["conflict_order_count"],
                    "conflict_rate": round(item["conflict_order_count"] / assigned_count, 4),
                    "replanned_order_count": item["replanned_order_count"],
                    "replan_rate": round(item["replanned_order_count"] / assigned_count, 4),
                    "avg_dispatch_response_minutes": self._average(item["response_minutes"]),
                }
            )
        items.sort(key=lambda value: (-int(value["order_count"]), str(value["group_key"])))
        return items

    async def get_performance_trend(
        self,
        *,
        window_start: datetime | None = None,
        window_end: datetime | None = None,
        bucket: str = "hour",
    ) -> list[dict[str, Any]]:
        if bucket != "hour":
            raise ValueError("仅支持按小时聚合趋势")

        ws, we = self._resolve_window(window_start, window_end)
        orders = await self._get_orders(ws, we)
        conflict_items = await self._get_conflicts(ws, we, len(orders))
        conflict_order_ids = self._conflict_order_ids(conflict_items)
        replan_order_ids = await self._get_replanned_order_ids(orders)

        buckets: dict[datetime, dict[str, Any]] = {}
        cursor = ws.replace(minute=0, second=0, microsecond=0)
        while cursor < we:
            buckets[cursor] = {
                "bucket_start": cursor,
                "bucket_end": cursor + timedelta(hours=1),
                "order_count": 0,
                "conflict_order_count": 0,
                "replanned_order_count": 0,
                "response_minutes": [],
            }
            cursor += timedelta(hours=1)

        for order in orders:
            start_time = self._order_start(order)
            if start_time is None:
                continue
            bucket_start = start_time.replace(minute=0, second=0, microsecond=0)
            if bucket_start not in buckets:
                continue
            item = buckets[bucket_start]
            order_id = str(getattr(order, "id", "")).strip()
            item["order_count"] += 1
            if order_id in conflict_order_ids:
                item["conflict_order_count"] += 1
            if order_id in replan_order_ids:
                item["replanned_order_count"] += 1
            response_minutes = self._response_minutes_for_order(order)
            if response_minutes is not None:
                item["response_minutes"].append(response_minutes)

        return [
            {
                "bucket_start": value["bucket_start"].isoformat(),
                "bucket_end": value["bucket_end"].isoformat(),
                "order_count": value["order_count"],
                "conflict_order_count": value["conflict_order_count"],
                "replanned_order_count": value["replanned_order_count"],
                "avg_dispatch_response_minutes": self._average(value["response_minutes"]),
            }
            for _, value in sorted(buckets.items(), key=lambda item: item[0])
        ]

    async def _get_orders(self, ws: datetime, we: datetime) -> list[Any]:
        return await self._order_repo.find_orders_in_window(
            window_start=ws,
            window_end=we,
            statuses=None,
        )

    async def _get_conflicts(self, ws: datetime, we: datetime, order_count: int) -> list[dict[str, Any]]:
        limit = max(200, max(1, order_count) * 8)
        return await self._conflict_service.list_conflicts(window_start=ws, window_end=we, limit=limit)

    async def _get_replanned_order_ids(self, orders: list[Any]) -> set[str]:
        if not hasattr(self._order_repo, "list_logs"):
            return set()

        result: set[str] = set()
        for order in orders:
            order_id = str(getattr(order, "id", "")).strip()
            if not order_id:
                continue
            logs = await self._order_repo.list_logs(order_id, limit=200)
            if any(self._is_replan_action(item.get("action")) for item in logs):
                result.add(order_id)
        return result

    @classmethod
    def _is_replan_action(cls, action: Any) -> bool:
        value = str(action or "").strip().lower()
        return any(value.startswith(prefix) for prefix in cls.REPLAN_ACTION_PREFIXES)

    @staticmethod
    def _conflict_order_ids(conflict_items: list[dict[str, Any]]) -> set[str]:
        result: set[str] = set()
        for item in conflict_items:
            for order_id in item.get("related_dispatch_order_ids") or []:
                value = str(order_id).strip()
                if value:
                    result.add(value)
        return result

    @staticmethod
    def _status_value(order: Any) -> str:
        status = getattr(order, "status", None)
        return str(getattr(status, "value", status) or "").strip().lower()

    @staticmethod
    def _is_assigned(order: Any) -> bool:
        return bool(
            getattr(order, "team_id", None)
            or getattr(order, "individual_user_id", None)
            or getattr(order, "members", None)
        )

    def _group_key(self, order: Any, group_by: str) -> tuple[str, str]:
        if group_by == "terminal":
            value = str(getattr(order, "terminal", None) or "unknown")
            return value, value
        if group_by == "step":
            key = str(getattr(order, "task_type", None) or "unknown")
            label = str(getattr(order, "task_type_name", None) or key)
            return key, label

        key = str(getattr(order, "team_id", None) or "unassigned")
        label = str(getattr(order, "team_name", None) or key)
        return key, label

    def _team_occupied_minutes(self, orders: list[Any], ws: datetime, we: datetime) -> dict[str, float]:
        team_loads: dict[str, float] = defaultdict(float)
        for order in orders:
            team_id = str(getattr(order, "team_id", "") or "").strip()
            if not team_id:
                continue
            team_loads[team_id] += self._occupied_minutes(order, ws, we)
        return dict(team_loads)

    def _response_minutes(self, orders: list[Any]) -> list[float]:
        return [value for value in (self._response_minutes_for_order(order) for order in orders) if value is not None]

    @staticmethod
    def _response_minutes_for_order(order: Any) -> float | None:
        created_at = getattr(order, "created_at", None)
        dispatched_at = getattr(order, "dispatched_at", None)
        if created_at is None or dispatched_at is None:
            return None
        diff = (dispatched_at - created_at).total_seconds() / 60.0
        if diff < 0:
            return None
        return round(diff, 2)

    @staticmethod
    def _occupied_minutes(order: Any, ws: datetime, we: datetime) -> float:
        start = DispatchAnalyticsService._order_start(order)
        end = DispatchAnalyticsService._order_end(order)
        if start is None:
            return 0.0
        if end is None:
            end = start + timedelta(minutes=30)
        overlap_start = max(start, ws)
        overlap_end = min(end, we)
        if overlap_end <= overlap_start:
            return 0.0
        return (overlap_end - overlap_start).total_seconds() / 60.0

    @staticmethod
    def _order_start(order: Any) -> datetime | None:
        return (
            getattr(order, "actual_start_time", None)
            or getattr(order, "planned_start_time", None)
            or getattr(order, "assignment_deadline", None)
            or getattr(order, "created_at", None)
        )

    @staticmethod
    def _order_end(order: Any) -> datetime | None:
        return (
            getattr(order, "actual_end_time", None)
            or getattr(order, "estimated_completion_time", None)
            or getattr(order, "planned_end_time", None)
            or getattr(order, "actual_start_time", None)
            or getattr(order, "planned_start_time", None)
        )

    def _key_order_stats(self, orders: list[Any]) -> dict[str, Any]:
        key_orders = [order for order in orders if self._is_key_order(order)]
        if not key_orders:
            return {"key_order_count": 0, "key_order_ontime_rate": 0.0}

        ontime_count = 0
        for order in key_orders:
            actual_end = self._order_end(order)
            planned_end = getattr(order, "planned_end_time", None)
            deadline = getattr(order, "assignment_deadline", None) or planned_end
            if actual_end is not None and deadline is not None and actual_end <= deadline:
                ontime_count += 1
        return {
            "key_order_count": len(key_orders),
            "key_order_ontime_rate": round(ontime_count / len(key_orders), 4),
        }

    @staticmethod
    def _is_key_order(order: Any) -> bool:
        workflow_context = getattr(order, "workflow_context", None) or {}
        if not isinstance(workflow_context, dict):
            return False
        if workflow_context.get("is_key_flight") is True:
            return True
        priority = (
            str(workflow_context.get("priority") or workflow_context.get("flight_priority") or "").strip().lower()
        )
        return priority in {"high", "critical", "urgent"}

    @staticmethod
    def _load_balance_score(loads: Iterable[float]) -> float:
        values = [float(item) for item in loads if float(item) >= 0]
        if not values:
            return 1.0
        if len(values) == 1:
            return 1.0
        mean = sum(values) / len(values)
        if mean <= 0:
            return 1.0
        variance = sum((value - mean) ** 2 for value in values) / len(values)
        coefficient = sqrt(variance) / mean
        return round(max(0.0, min(1.0, 1.0 - coefficient)), 4)

    @staticmethod
    def _average(values: Iterable[float]) -> float:
        normalized = [float(item) for item in values]
        if not normalized:
            return 0.0
        return round(sum(normalized) / len(normalized), 2)

    @staticmethod
    def _resolve_window(
        window_start: datetime | None,
        window_end: datetime | None,
    ) -> tuple[datetime, datetime]:
        now = utc_now()
        end = window_end or now
        start = window_start or (end - timedelta(hours=12))
        return start, end

File: services/api-server/crates/application/src/services/dispatch_analytics_service.rs
//! 派工运营分析服务。

use chrono::{DateTime, Duration, Timelike, Utc};
use serde_json::Value;
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

use crate::schemas::dispatch_schemas::{
    DispatchAnalyticsBreakdownItem, DispatchAnalyticsSummaryResponse, DispatchAnalyticsTrendItem,
};
use crate::types::{ConcreteDispatchQueryService, ConcreteResourceUtilizationService};
use fms_domain::error::DomainError;
use fms_domain::models::dispatch::DispatchOrder;
use fms_domain::ports::dispatch_repository::DispatchOrderRepository;

pub struct DispatchAnalyticsService {
    order_repo: Arc<dyn DispatchOrderRepository + Send + Sync>,
    query_service: Arc<ConcreteDispatchQueryService>,
    resource_utilization_service: Arc<ConcreteResourceUtilizationService>,
}

impl DispatchAnalyticsService {
    pub fn new(
        order_repo: Arc<dyn DispatchOrderRepository + Send + Sync>,
        query_service: Arc<ConcreteDispatchQueryService>,
        resource_utilization_service: Arc<ConcreteResourceUtilizationService>,
    ) -> Self {
        Self {
            order_repo,
            query_service,
            resource_utilization_service,
        }
    }

    pub async fn get_operations_summary(
        &self,
        window_start: Option<DateTime<Utc>>,
        window_end: Option<DateTime<Utc>>,
    ) -> Result<DispatchAnalyticsSummaryResponse, DomainError> {
        let (ws, we) = resolve_window(window_start, window_end);
        let orders = self.get_orders(ws, we).await?;
        let conflict_items = self.get_conflicts(ws, we, orders.len()).await?;
        let conflict_order_ids = conflict_order_ids(&conflict_items);
        let replanned_order_ids = self.get_replanned_order_ids(&orders).await?;
        let response_minutes = orders.iter().filter_map(response_minutes_for_order).collect::<Vec<_>>();
        let team_loads = team_occupied_minutes(&orders, ws, we);
        let equipment_utilization = self
            .resource_utilization_service
            .get_equipment_utilization(Some(ws), Some(we))
            .await?;
        let key_stats = key_order_stats(&orders);
        let assigned_orders = orders.iter().filter(|order| is_assigned(order)).count() as i32;
        let completed_orders = orders.iter().filter(|order| status_value(order) == "completed").count() as i32;

        let equipment_idle_rate = if equipment_utilization.is_empty() {
            0.0
        } else {
            let avg_rate = equipment_utilization
                .iter()
                .filter_map(|item| item.get("utilization_rate").and_then(Value::as_f64))
                .sum::<f64>()
                / equipment_utilization.len() as f64;
            round_to_4((1.0_f64 - avg_rate).max(0.0_f64))
        };
        let denominator = assigned_orders.max(1) as f64;

        Ok(DispatchAnalyticsSummaryResponse {
            window_start: ws,
            window_end: we,
            assigned_order_count: assigned_orders,
            completed_order_count: completed_orders,
            conflict_count: conflict_items.len() as i32,
            conflict_order_count: conflict_order_ids.len() as i32,
            conflict_rate: round_to_4(conflict_order_ids.len() as f64 / denominator),
            replanned_order_count: replanned_order_ids.len() as i32,
            replan_rate: round_to_4(replanned_order_ids.len() as f64 / denominator),
            avg_dispatch_response_minutes: average(&response_minutes),
            team_load_balance_score: load_balance_score(team_loads.values().copied().collect()),
            equipment_idle_rate,
            key_order_count: key_stats.0,
            key_order_ontime_rate: key_stats.1,
        })
    }

    pub async fn get_breakdown(
        &self,
        window_start: Option<DateTime<Utc>>,
        window_end: Option<DateTime<Utc>>,
        group_by: &str,
    ) -> Result<Vec<DispatchAnalyticsBreakdownItem>, DomainError> {
        let (ws, we) = resolve_window(window_start, window_end);
        let orders = self.get_orders(ws, we).await?;
        let conflict_items = self.get_conflicts(ws, we, orders.len()).await?;
        let conflict_order_ids = conflict_order_ids(&conflict_items);
        let replanned_order_ids = self.get_replanned_order_ids(&orders).await?;

        #[derive(Default)]
        struct GroupStats {
            group_label: String,
            order_count: i32,
            assigned_order_count: i32,
            completed_order_count: i32,
            occupied_minutes: f64,
            conflict_order_count: i32,
            replanned_order_count: i32,
            response_minutes: Vec<f64>,
        }

        let mut groups: HashMap<String, GroupStats> = HashMap::new();
        for order in &orders {
            let (key, label) = group_key(order, group_by);
            let entry = groups.entry(key.clone()).or_default();
            entry.group_label = label;
            entry.order_count += 1;
            if is_assigned(order) {
                entry.assigned_order_count += 1;
            }
            if status_value(order) == "completed" {
                entry.completed_order_count += 1;
            }
            entry.occupied_minutes += occupied_minutes(order, ws, we);
            if conflict_order_ids.contains(&order.id) {
                entry.conflict_order_count += 1;
            }
            if replanned_order_ids.contains(&order.id) {
                entry.replanned_order_count += 1;
            }
            if let Some(value) = response_minutes_for_order(order) {
                entry.response_minutes.push(value);
            }
        }

        let mut items = groups
            .into_iter()
            .map(|(group_key, stats)| {
                let assigned_count = stats.assigned_order_count.max(1) as f64;
                DispatchAnalyticsBreakdownItem {
                    group_key,
                    group_label: stats.group_label,
                    order_count: stats.order_count,
                    assigned_order_count: stats.assigned_order_count,
                    completed_order_count: stats.completed_order_count,
                    occupied_minutes: round_to_2(stats.occupied_minutes),
                    conflict_order_count: stats.conflict_order_count,
                    conflict_rate: round_to_4(stats.conflict_order_count as f64 / assigned_count),
                    replanned_order_count: stats.replanned_order_count,
                    replan_rate: round_to_4(stats.replanned_order_count as f64 / assigned_count),
                    avg_dispatch_response_minutes: average(&stats.response_minutes),
                }
            })
            .collect::<Vec<_>>();
        items.sort_by(|left, right| {
            right
                .order_count
                .cmp(&left.order_count)
                .then_with(|| left.group_key.cmp(&right.group_key))
        });
        Ok(items)
    }

    pub async fn get_performance_trend(
        &self,
        window_start: Option<DateTime<Utc>>,
        window_end: Option<DateTime<Utc>>,
        bucket: &str,
    ) -> Result<Vec<DispatchAnalyticsTrendItem>, DomainError> {
        if bucket != "hour" {
            return Err(DomainError::ValidationError("仅支持按小时聚合趋势".to_string()));
        }
        let (ws, we) = resolve_window(window_start, window_end);
        let orders = self.get_orders(ws, we).await?;
        let conflict_items = self.get_conflicts(ws, we, orders.len()).await?;
        let conflict_order_ids = conflict_order_ids(&conflict_items);
        let replanned_order_ids = self.get_replanned_order_ids(&orders).await?;

        #[derive(Default)]
        struct BucketStats {
            order_count: i32,
            conflict_order_count: i32,
            replanned_order_count: i32,
            response_minutes: Vec<f64>,
        }

        let mut buckets: HashMap<DateTime<Utc>, BucketStats> = HashMap::new();
        let mut cursor = ws
            .with_minute(0)
            .and_then(|value| value.with_second(0))
            .and_then(|value| value.with_nanosecond(0))
            .unwrap_or(ws);
        while cursor < we {
            buckets.insert(cursor, BucketStats::default());
            cursor += Duration::hours(1);
        }

        for order in &orders {
            let Some(start_time) = order_start(order) else {
                continue;
            };
            let Some(bucket_start) = start_time
                .with_minute(0)
                .and_then(|value| value.with_second(0))
                .and_then(|value| value.with_nanosecond(0))
            else {
                continue;
            };
            let Some(bucket_stats) = buckets.get_mut(&bucket_start) else {
                continue;
            };
            bucket_stats.order_count += 1;
            if conflict_order_ids.contains(&order.id) {
                bucket_stats.conflict_order_count += 1;
            }
            if replanned_order_ids.contains(&order.id) {
                bucket_stats.replanned_order_count += 1;
            }
            if let Some(value) = response_minutes_for_order(order) {
                bucket_stats.response_minutes.push(value);
            }
        }

        let mut items = buckets.into_iter().collect::<Vec<_>>();
        items.sort_by_key(|(bucket_start, _)| *bucket_start);
        Ok(items
            .into_iter()
            .map(|(bucket_start, stats)| DispatchAnalyticsTrendItem {
                bucket_start,
                bucket_end: bucket_start + Duration::hours(1),
                order_count: stats.order_count,
                conflict_order_count: stats.conflict_order_count,
                replanned_order_count: stats.replanned_order_count,
                avg_dispatch_response_minutes: average(&stats.response_minutes),
            })
            .collect())
    }

    async fn get_orders(&self, ws: DateTime<Utc>, we: DateTime<Utc>) -> Result<Vec<DispatchOrder>, DomainError> {
        self.order_repo
            .find_orders_in_window(ws, we, &[], None, None, None, false)
            .await
    }

    async fn get_conflicts(
        &self,
        ws: DateTime<Utc>,
        we: DateTime<Utc>,
        order_count: usize,
    ) -> Result<Vec<Value>, DomainError> {
        let limit = (order_count.max(1) * 8).max(200) as i64;
        self.query_service.list_conflicts_for_analytics(ws, we, limit).await
    }

    async fn get_replanned_order_ids(&self, orders: &[DispatchOrder]) -> Result<HashSet<String>, DomainError> {
        let mut result = HashSet::new();
        for order in orders {
            let logs = self.order_repo.list_logs(&order.id, 200).await?;
            if logs.iter().any(|item| {
                item.get("action")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_ascii_lowercase()
                    .starts_with("replanned")
            }) {
                result.insert(order.id.clone());
            }
        }
        Ok(result)
    }
}

fn resolve_window(
    window_start: Option<DateTime<Utc>>,
    window_end: Option<DateTime<Utc>>,
) -> (DateTime<Utc>, DateTime<Utc>) {
    let end = window_end.unwrap_or_else(Utc::now);
    let start = window_start.unwrap_or_else(|| end - Duration::hours(12));
    (start, end)
}

fn conflict_order_ids(items: &[Value]) -> HashSet<String> {
    let mut result = HashSet::new();
    for item in items {
        if let Some(order_ids) = item.get("related_dispatch_order_ids").and_then(Value::as_array) {
            for order_id in order_ids {
                if let Some(value) = order_id.as_str() {
                    result.insert(value.to_string());
                }
            }
        }
    }
    result
}

fn status_value(order: &DispatchOrder) -> &'static str {
    match order.status {
        fms_domain::models::dispatch::DispatchOrderStatus::Pending => "pending",
        fms_domain::models::dispatch::DispatchOrderStatus::Assigned => "assigned",
        fms_domain::models::dispatch::DispatchOrderStatus::InProgress => "in_progress",
        fms_domain::models::dispatch::DispatchOrderStatus::Completed => "completed",
        fms_domain::models::dispatch::DispatchOrderStatus::Cancelled => "cancelled",
    }
}

fn is_assigned(order: &DispatchOrder) -> bool {
    order.individual_user_id.is_some() || !order.members.is_empty()
}

fn group_key(order: &DispatchOrder, group_by: &str) -> (String, String) {
    match group_by {
        "terminal" => {
            let value = order.terminal.clone().unwrap_or_else(|| "unknown".to_string());
            (value.clone(), value)
        }
        "step" => {
            let key = order.task_type.clone();
            let label = order.task_type_name.clone().unwrap_or_else(|| key.clone());
            (key, label)
        }
        // 班组不再是指派对象：默认按科室分组
        _ => {
            let key = order
                .department_id
                .clone()
                .or_else(|| order.department.clone())
                .unwrap_or_else(|| "unassigned".to_string());
            (key.clone(), key)
        }
    }
}

fn team_occupied_minutes(orders: &[DispatchOrder], ws: DateTime<Utc>, we: DateTime<Utc>) -> HashMap<String, f64> {
    // 工单不再挂班组：按科室聚合占用时长
    let mut result = HashMap::new();
    for order in orders {
        let Some(department_id) = order.department_id.as_ref() else {
            continue;
        };
        *result.entry(department_id.clone()).or_insert(0.0) += occupied_minutes(order, ws, we);
    }
    result
}

fn response_minutes_for_order(order: &DispatchOrder) -> Option<f64> {
    let created_at = order.created_at?;
    let dispatched_at = order.dispatched_at?;
    let diff = (dispatched_at - created_at).num_seconds() as f64 / 60.0;
    if diff < 0.0 {
        None
    } else {
        Some(round_to_2(diff))
    }
}

fn occupied_minutes(order: &DispatchOrder, ws: DateTime<Utc>, we: DateTime<Utc>) -> f64 {
    let Some(start) = order_start(order) else {
        return 0.0;
    };
    let end = order_end(order).unwrap_or(start + Duration::minutes(30));
    let overlap_start = start.max(ws);
    let overlap_end = end.min(we);
    if overlap_end <= overlap_start {
        return 0.0;
    }
    (overlap_end - overlap_start).num_seconds() as f64 / 60.0
}

fn order_start(order: &DispatchOrder) -> Option<DateTime<Utc>> {
    order
        .actual_start_time
        .or(order.planned_start_time)
        .or(order.assignment_deadline)
        .or(order.created_at)
}

fn order_end(order: &DispatchOrder) -> Option<DateTime<Utc>> {
    order
        .actual_end_time
        .or(order.estimated_completion_time)
        .or(order.planned_end_time)
        .or(order.actual_start_time)
        .or(order.planned_start_time)
}

fn key_order_stats(orders: &[DispatchOrder]) -> (i32, f64) {
    let key_orders = orders.iter().filter(|order| is_key_order(order)).collect::<Vec<_>>();
    if key_orders.is_empty() {
        return (0, 0.0);
    }
    let on_time_count = key_orders
        .iter()
        .filter(|order| {
            let actual_end = order_end(order);
            let deadline = order.assignment_deadline.or(order.planned_end_time);
            matches!((actual_end, deadline), (Some(actual_end), Some(deadline)) if actual_end <= deadline)
        })
        .count() as i32;
    (
        key_orders.len() as i32,
        round_to_4(on_time_count as f64 / key_orders.len() as f64),
    )
}

fn is_key_order(order: &DispatchOrder) -> bool {
    if order
        .workflow_context
        .get("is_key_flight")
        .and_then(Value::as_bool)
        .unwrap_or(false)
    {
        return true;
    }
    let priority = order
        .workflow_context
        .get("priority")
        .or_else(|| order.workflow_context.get("flight_priority"))
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_ascii_lowercase();
    matches!(priority.as_str(), "high" | "critical" | "urgent")
}

fn load_balance_score(values: Vec<f64>) -> f64 {
    if values.len() <= 1 {
        return 1.0;
    }
    let mean = values.iter().sum::<f64>() / values.len() as f64;
    if mean <= 0.0 {
        return 1.0;
    }
    let variance = values.iter().map(|value| (*value - mean).powi(2)).sum::<f64>() / values.len() as f64;
    let coefficient = variance.sqrt() / mean;
    round_to_4((1.0 - coefficient).clamp(0.0, 1.0))
}

fn average(values: &[f64]) -> f64 {
    if values.is_empty() {
        0.0
    } else {
        round_to_2(values.iter().sum::<f64>() / values.len() as f64)
    }
}

fn round_to_2(value: f64) -> f64 {
    (value * 100.0).round() / 100.0
}

fn round_to_4(value: f64) -> f64 {
    (value * 10_000.0).round() / 10_000.0
}
